### ISSUE TITLE
Are there any plans to use NMOS in the future?

### ISSUE THREAD

**Issue by calanor** [bondagit/aes67-linux-daemon#113]
It would be very useful to be able to use NMOS IS-04 for discovery and IS-05 for automatic sink configuration.
The libraries published by Sony could also be useful:
https://github.com/sony/nmos-cpp

**Comment by andyweiss** [comment]
This would be very useful for me as well! That‘s the missing piece for the use in orchestrated environments!

**Comment by br-online** [comment]
Same here... NMOS would be great!Am 09.02.2023 14:40 schrieb andyweiss ***@***.***>:

This would be very useful for me as well! That‘s the missing piece for the use in orchestrated environments!

**Comment by bondagit** [comment]
yes, it's certainly possible to integrate NMOS IS-04 node functions into the daemon.
Are you aware of any cheap AES67 device implementing NMOS IS-04 to run some interoperability tests ?

**Comment by andyweiss** [comment]
It will be hard to find any cheap device capable of NMOS IS-04. There are some option to test against other software.
I can collect some links if you are interested. I have acces to some AES67 NMOS capaple devices and can offer to do various tests.

**Comment by calanor** [comment]
In nmos-cpp API from Sony have som exemples and a code for make a NMOS node.  Could use  as a device.

**Comment by bondagit** [comment]
ok, thanks for the info. I will be working on this when possible.

**Comment by stvmyr** [comment]
im doing a lot of research an testing into SMPTE ST 2110 lately. An NMOS implementation would be great for using your daemon in such environment! Maybe the NMOS IS-08 standard for Audio channel mapping could be interesting too.

if you still looking for a interoperability test, you could have a look into this:
https://github.com/rhastie/easy-nmos

easy-nmos comes with testing tools, registry, controller and sample node.

You could also find the standalone testing tool here:
https://github.com/AMWA-TV/nmos-testing

**Comment by tellierflexus** [comment]
Hello ! Any update on this ? I could help test NMOS with pleasure if you need to.

**Comment by D21Greg** [comment]
Hi everybody, I did (not I but Claude) a functioning version of added IS-04 and IS-05 nmos update for the daemon. So as I have no idea how to reference the remarkable work of M. Bondovalli, I apologize if not published correctly an am totally open to change the credits of this : https://github.com/GBA-TAB/nmos-aes67-daemon . Maybe the functionality could be added to the master fork. So what I still have to test is 2022/7 functionality and validate DNS-SD for register discovery. It has been tested on my minimal home bench (so not 2022/7) BMD 3x3 and Decklink with fix pointed register.  Not yet fully documented but all the config stuff is in daemon.conf and now on the UI. Let me know what the result of your tests are. It was built against testing tool and passed everything. BCP 008 is the next, and lets see how to shred IS-08 :).

**Comment by bondagit** [comment]
HI @D21Greg , thanks for the great contribution ! 
I created a new branch named [mos-aes67-daemon](https://github.com/bondagit/aes67-linux-daemon/tree/nmos-aes67-daemon) to integrate your contribution and will run some tests using [easy-nmos](https://github.com/rhastie/easy-nmos).

**Comment by D21Greg** [comment]
Hi Andrea,
let me know if you need some BMD stuff, I can lend you some if you need. So copper 10 Gig SDI 3x3, decklink IP copper or SMF SFP+ (very bad firmware :) ) . 
If you need Riedel NMOS explorer, here a link for downloading it :
https://www.dropbox.com/scl/fo/rglvostkvqz0h3tkzhgvk/AKzUDkdLlPheUGjbIG32gNQ?rlkey=stwvwslccvklmmrhz7eqd8q0s&st=tqyo5f1v&dl=0￼
NMOS
dropbox.com

**Comment by stvmyr** [comment]
great to see NMOS in the project!
We got a Phabrix QxL to verify 2022-7 aswell as multiple devices that supports sending and receiving it. 

I will certainly give it a try myself In the next days!

**Comment by LeeO86** [comment]
That’s really interesting – this comes up at exactly the same time I was working on an additional service that maps NMOS to the REST API.

You can try it here: [aes67-nmos-bridge](https://github.com/LeeO86/aes67-nmos-bridge)

Initially, I thought implementing NMOS directly into the daemon would be beyond my skills, and I also wasn’t confident I could submit a PR that would meet the project’s quality standards. The last thing I wanted was to contribute an AI-generated drive-by PR.

After developing the bridge, I actually think integrating NMOS directly into the daemon is the better approach. There is quite a bit of configuration that has to stay in sync between the daemon and the bridge, and the bridge also has to poll the REST API regularly, which feels less than ideal.

I’ve tested my implementation successfully with a Merging Anubis and a DirectOut ExBox.RAV. For easier testing, I’ve been using this fork of NMOS Crosspoint: [NMOS-Crosspoint Fork](https://github.com/Gemini2350/nmos-crosspoint_v3).

**Comment by bondagit** [comment]
Thanks for your interest.

I recently set up a simple NMOS test environment using a registry from nmos-cpp together with the Riedel NMOS Explorer. I was able to successfully register and discover a couple of AES67 nodes (see the attached snapshot).
<img width="1540" height="1110" alt="Image" src="https://github.com/user-attachments/assets/cd28d887-11fd-4506-ae9c-6b65b4dd026b" />

Along the way, I made a few changes that might be useful for everyone:

- The AES67 daemon now updates its node registration in the registry whenever the PTP clock status changes.
- The node registration uses the default node ID, making it easier to distinguish between multiple daemon instances.
- Added an NMOS Configuration section to the Web UI, including the enable flag, registry address and port, and node port.
- Updated the documentation. I would also like to document the steps required to recreate the simple test setup I am currently using.

Do the recent changes make sense to you ? 
Since I am not an NMOS expert, I would appreciate your help in identifying any missing features, potential compatibility issues, and other areas for improvement. Contributions and feedback are very welcome.

**Comment by D21Greg** [comment]
Nice,
this is my point of view, it can differ from mainline way of thinking. 
DNS-SD (so for registration (auto discovery of the register by the node)) is appreciated for small systems, then for bigger systems (or routed ones), all what is link local based should be eliminated to limit traffic. So the option is appreciated, could be on the GUI.
Advertising Node thru DNS-SD has the same logic, so very practical for register-less systems, not practical for bigger ones.
Registration names (Node/device/ressources) must exist :) and very good if correlated to ie hostnames ? or free editable ?
having UI controlling that is good.
I test it ASAP (Sorry, lots of festivals setups to work on, so free time rules :) )

The objective is orchestration, I look for to drop a simple orchestrator NMOS ASAP, so to have an open as possible non dependant base for orchestration. Another path is to 

Actually, IS-04 and IS-05 are widely available. IS-12 is coming (or should :) ). IS-08 is not scheduled, but cool to have a test-bench :).

We are not so concerned with BCP-002 on its actual form, as it is an audio only device. 

Dash 7 is a good discussion, it should be actually implemented with primary transport, secondary transport, and OOB, so 3 interfaces, with all the options to present NMOS on only one (traditionally OOB), or all the interfaces, so there is are lots of possibilities, as big systems are traditionnaly built on the 3 interfaces logic, with a tag for where NMOS is presented. The integration of 1 interface nodes is mostly solved on routing level. So maybe options for services per interface ? 

So I continue creative/conceptual prototypes  and keep you up to date as soon as I hit something important. 

Thanks for the fantastic work achieved with this project.

**Comment by bondagit** [comment]
NMOS support is now merged into daemon v4.0.0.
In case of problems please open separate  issues.
